### PR TITLE
feat: enable SecureServing and TLS profile on metrics endpoint

### PR DESCRIPTION
**Pull request by ugiordan**
## Summary

Adds OpenShift TLS profile integration to the batch-gateway operator: SecureServing on the operator metrics endpoint, TLS adherence policy support, SecurityProfileWatcher for graceful restart on profile changes, and injection of the cluster TLS profile into the batch-gateway workload deployment.

## Changes

- **SecureServing on metrics**: `metricsserver.Options{SecureServing: true}` with cluster TLS profile (MinVersion, CipherSuites, ALPN h2/http1.1)
- **TLS adherence policy**: Fetches `TLSAdherencePolicy` from APIServer CR at startup
- **SecurityProfileWatcher**: Watches for TLS profile and adherence policy changes, triggers graceful restart
- **Workload TLS injection**: Passes cluster TLS profile values to batch-gateway helm chart (`apiserver.tls.minVersion`, `cipherSuites`, `nextProtos`) via `TLSProfileValues` struct
- **Batch-gateway ref bump**: `BATCH_GATEWAY_REF` updated to `c863b1e` (includes TLS CLI flags from upstream [llm-d/llm-d-batch-gateway#618](https://github.com/llm-d/llm-d-batch-gateway/pull/618), merged)
- **service-ca cert provisioning**: Annotation on metrics Service, cert mounted at `/tmp/k8s-metrics-server/metrics-certs`
- **ServiceMonitor**: Updated for HTTPS scraping with `caFile` and `serverName`
- **RBAC**: `config.openshift.io/apiservers` with `get`, `list`, `watch` (list+watch for SecurityProfileWatcher)
- **Transient error handling**: ServiceUnavailable, Timeout, ServerTimeout, TooManyRequests, DeadlineExceeded fall back to Intermediate defaults with watcher self-healing

## Why one PR

The operator doesn't currently depend on `openshift/api` or `controller-runtime-common`. Adding these dependencies (required for TLS profile reading) pulls in Go 1.26 and controller-runtime v0.24.1. Splitting into two PRs would require the same dependency cascade in both, so combining them is cleaner.

## Error handling

- IsNoMatchError: graceful fallback (non-OpenShift cluster)
- IsNotFound: graceful fallback
- IsForbidden / unexpected: fail-closed (os.Exit)
- Transient errors: log + Intermediate fallback + register watcher for self-healing

## Jira

- [RHOAIENG-78721](https://redhat.atlassian.net/browse/RHOAIENG-78721): SecureServing on metrics
- [RHOAIENG-78722](https://redhat.atlassian.net/browse/RHOAIENG-78722): Workload TLS profile injection
- [RHOAIENG-78973](https://redhat.atlassian.net/browse/RHOAIENG-78973): SecurityProfileWatcher + TLSAdherencePolicy

## Testing

- `go build ./...` passes
- All existing tests updated for new `NewHelmRenderer` signature

[RHOAIENG-78721]: https://redhat.atlassian.net/browse/RHOAIENG-78721?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -106,10 +117,52 @@ func main() {
...
+
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme: scheme,
 		Metrics: metricsserver.Options{
-			BindAddress: metricsAddr,
+			BindAddress:    metricsAddr,
+			SecureServing:  true,
+			TLSOpts:        tlsOpts,
+			FilterProvider: filters.WithAuthenticationAndAuthorization,
 		},
```
[the remark this answers is not in the document]
Comment (ugiordan) [reply cmd/main.go]: @coderabbitai could you verify again? I've fixed this one.

File: .github/workflows/ci-image.yml [hunk 2]
```diff
@@ -23,8 +23,6 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
-        with:
```
Comment (yizhaodev) [inline .github/workflows/ci-image.yml:26]: why thoese are removed?

Comment (ugiordan) [reply .github/workflows/ci-image.yml:26]: The `persist-credentials: false` removal was unintentional (lost during squash). It's restored now in the latest push.

File: cmd/main.go [hunk 3]
```diff
@@ -106,10 +116,69 @@ func main() {
...
+	restConfig := ctrl.GetConfigOrDie()
+	bootstrapClient, err := client.New(restConfig, client.Options{Scheme: scheme})
+	if err != nil {
+		logger.Error(err, "unable to create bootstrap client")
+		os.Exit(1)
+	}
+
+	var tlsOpts []func(*tls.Config)
+
+	bootstrapCtx, cancelBootstrap := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (acardace) [inline cmd/main.go]: why don't you just `defer cancelBootstrap()` instead of calling it in the default branch? It would make the code cleaner.

File: cmd/main.go [hunk 4]
```diff
@@ -106,10 +116,69 @@ func main() {
 	logger := klog.NewKlogr()
 	ctrl.SetLogger(logger)
 
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+	// Resolve the cluster TLS profile for secure metrics serving.
+	restConfig := ctrl.GetConfigOrDie()
+	bootstrapClient, err := client.New(restConfig, client.Options{Scheme: scheme})
```
Comment (acardace) [inline cmd/main.go:118]: can you move all of this init logic to a new file inside the monitoring package?

File: cmd/main.go [hunk 5]
```diff
@@ -173,8 +244,33 @@ func main() {
 		logger.Info("POD_NAMESPACE not set, skipping metrics controller reconciliation")
 	}
 
+	ctx, cancel := context.WithCancel(ctrl.SetupSignalHandler())
```
Comment (acardace) [inline cmd/main.go:203]: Again I'd expect to see all this logic to be moved inside its own file with clear unit testable functions.

File: cmd/main.go [hunk 6]
```diff
@@ -106,10 +116,69 @@ func main() {
...
+			apierrors.IsTooManyRequests(err),
+			errors.Is(err, context.DeadlineExceeded):
+			logger.Info("Transient API error, using Intermediate defaults", "error", err)
+			tlsProfileFetched = true
+		default:
+			cancelBootstrap()
+			logger.Error(err, "unable to read TLS profile")
+			os.Exit(1)
+		}
+		tlsProfile = *configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
```
Comment (acardace) [inline cmd/main.go]: does this work correctly on vanilla kubernetes?

**Review by acardace — CHANGES_REQUESTED** [review]
The cluster TLS profile is injected directly into helm values at HelmRenderer construction time, bypassing the CRD entirely. This has a few issues:

- The TLSProfileValues is stored on the HelmRenderer struct and passed to both renderers, but only the batch renderer uses it, the async renderer ignores it.
- Users have no visibility into or control over these values since they don't appear on the CR.
- It's inconsistent with how every other config value flows through the system (CR spec -> helm values).

I think the right approach is to add minVersion, cipherSuites, nextProtos fields to the TLS spec in the CRD, and default them at runtime (either in a mutating webhook or in the reconciler) when unset by reading the cluster TLS profile. That way both renderers just read from the CR like they do for everything else, and users can inspect/override the values if needed.

File: cmd/main.go (hunk 3 above)
Comment (ugiordan) [reply cmd/main.go]: Done. `Resolve()` in `internal/tls/profile.go` now handles its own context with `defer cancel()` internally, so main.go never touches the bootstrap context directly. No more `exitAfterDefer` risk.

File: cmd/main.go (hunk 4 above)
Comment (ugiordan) [reply cmd/main.go:118]: Done. All TLS init logic is now in `internal/tls/profile.go` with unit-testable functions:
- `Resolve()`: fetches TLS profile + adherence, handles error classification, returns a `Result`
- `SetupWatcher()`: registers SecurityProfileWatcher conditionally
- `ProfileValuesFromSpec()`: converts spec to helm chart values

main.go just calls `tlsinit.Resolve()` and `tlsinit.SetupWatcher()` (3 lines total instead of ~75).

I put it in `internal/tls` rather than `internal/monitoring` since TLS profile resolution is used by both the metrics server config and the helm renderer, not just monitoring.

File: cmd/main.go (hunk 5 above)
Comment (ugiordan) [reply cmd/main.go:203]: Done. `SetupWatcher()` is in `internal/tls/profile.go`. main.go is now just:

    if err := tlsinit.SetupWatcher(mgr, tlsResult, cancel, logger); err != nil {

Unit tests added for `Resolve()` covering: APIServer not found (Intermediate fallback), custom profile (Modern), and TLS 1.3 no-ciphers edge case.

File: cmd/main.go (hunk 6 above)
Comment (ugiordan) [reply cmd/main.go]: Yes. `Resolve()` handles `IsNoMatchError` (CRD not registered on vanilla k8s) gracefully:
- Falls back to Intermediate defaults
- Sets `ProfileFetched = false`, so `SetupWatcher()` is a no-op (no watcher registered)
- The RBAC for `config.openshift.io` is in a separate kustomize overlay (`config/overlays/odh/rbac/tls/`), so vanilla clusters don't get those ClusterRole rules

Same pattern as ai-gateway-operator #76.

File: config/rbac/role.yaml [hunk 7]
```diff
@@ -2,7 +2,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
-  name: operator
+  name: manager-role
```
Comment (aneeshkp) [inline config/rbac/role.yaml]: The ClusterRole name changed from operator to manager-role, but config/rbac/role_binding.yaml still has roleRef.name: operator.

Comment (ugiordan) [reply config/rbac/role.yaml]: Good catch. Fixed and regenerated with the correct roleName=operator from the Makefile.

**Review by acardace — APPROVED** [review]
